File: leetcode/python3/median-of-two-sorted-arrays.py
#
# @lc app=leetcode id=4 lang=python3
#
# [4] Median of Two Sorted Arrays
#
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        l = len(nums1) + len(nums2)
        if l % 2 == 1:
            return self.findKth(nums1, nums2, l//2+1)
        else:
            return (self.findKth(nums1, nums2, l//2) + self.findKth(nums1, nums2, l//2+1)) / 2
        
    def findKth(self, A, B, k):
        if not A:
            return B[k-1]
        if not B:
            return A[k-1]
        if k == 1:
            return min(A[0], B[0])

        a = A[k//2-1] if len(A) >= k//2 else None
        b = B[k//2-1] if len(B) >= k//2 else None

        if b is None or (a is not None and a < b):
            return self.findKth(A[k//2:], B, k-k//2)
        else:
            return self.findKth(A, B[k//2:], k-k//2)
            

File: leetcode/cpp/median-of-two-sorted-arrays.cpp
/*
 * @lc app=leetcode id=4 lang=cpp
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size()) {
            return findMedianSortedArrays(nums2, nums1);
        }
        int x = nums1.size(), y = nums2.size();
        int l = 0, r = x;
        while (l <= r) {
            int partitionX = l + r >> 1;
            int partitionY = (x + y + 1) / 2 - partitionX;

            int maxLeftX = partitionX <= 0 ? INT_MIN : nums1[partitionX-1];
            int minRightX = partitionX >= x ? INT_MAX : nums1[partitionX];

            int maxLeftY = partitionY <= 0 ? INT_MIN : nums2[partitionY-1];
            int minRightY = partitionY >= y ? INT_MAX : nums2[partitionY];

            if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
                if ((x+y) % 2 == 0)
                    return (double)(max(maxLeftX, maxLeftY) + min(minRightX, minRightY)) / 2.0;
                else
                    return (double)max(maxLeftX, maxLeftY);
            } else if (maxLeftX > minRightY) {
                r = partitionX - 1;
            } else {
                l = partitionX + 1;
            }
        }
        return -1;
    }
};
// @lc code=end

